bokehjs/test/unit/core/util/spatial.ts

```diff
@@ -4,16 +4,119 @@ import {SpatialIndex} from "@bokehjs/core/util/spatial"
 import * as bbox from "@bokehjs/core/util/bbox"
 
 describe("core/util/spatial module", () => {
-  it("support SpatialIndex.bounds()", () => {
+
+  it("support SpatialIndex.bounds() with an empty index", () => {
+    const index = new SpatialIndex(0)
+    index.finish()
+    expect(index.bounds(bbox.empty())).to.be.equal(bbox.empty())
+    expect(index.bounds(bbox.full())).to.be.equal(bbox.empty())
+    expect(index.bounds(bbox.positive_x())).to.be.equal(bbox.empty())
+    expect(index.bounds(bbox.positive_y())).to.be.equal(bbox.empty())
+    expect(index.bounds(bbox.negative_x())).to.be.equal(bbox.empty())
+    expect(index.bounds(bbox.negative_y())).to.be.equal(bbox.empty())
+    expect(index.bounds(bbox.x_range(-1.5, 2.8))).to.be.equal(bbox.empty())
+    expect(index.bounds(bbox.y_range(-1.5, 2.8))).to.be.equal(bbox.empty())
+    expect(index.bounds({x0: -10, y0: 2.3, x1: 1000.2, y1: 100})).to.be.equal(bbox.empty())
+  })
+
+  it("support SpatialIndex.bounds() with full rect", () => {
+    const index = new SpatialIndex(6)
+    index.add_rect(0, 1, 1, 2)
+    index.add_rect(0.0001, 1, 1, 2)
+    index.add_rect(0, 100, 1, 200)
+    index.add_rect(0.0001, -100, 1, 200)
+    index.add_rect(0, 10, 1, 20)
+    index.add_rect(-0.0001, 10, 1, 20)
+    index.finish()
+    expect(index.bounds(bbox.full())).to.be.equal({x0: -0.0001, x1: 1, y0: -100, y1: 200})
+  })
+
+  // it("support SpatialIndex.bounds() with empty rect", () => {
+  //   const index = new SpatialIndex(6)
+  //   index.add_rect(0, 1, 1, 2)
+  //   index.add_rect(0.0001, 1, 1, 2)
+  //   index.add_rect(0, 100, 1, 200)
+  //   index.add_rect(0.0001, -100, 1, 200)
+  //   index.add_rect(0, 10, 1, 20)
+  //   index.add_rect(-0.0001, 10, 1, 20)
+  //   index.finish()
+  //   expect(index.bounds(bbox.empty())).to.be.equal(bbox.empty())
+  // })
```

Comment:
This test fails and I don't quite understand why yet

Reply:
@mattpap do you agree that searching the index with an empty bbox should return empty restults (i.e. no indices or empty bounds)? I want to make sure I am not misunderstanding what and "empty" bbox is intended to represent.

Reply:
Oh, this is failing because the call to `this._normalize(rect)` turns `empty()` into `full()` @mattpap is this what you intended? Presumably `_normalize` could be made to operate on finite values, but should it?

Reply:
```diff
diff --git a/bokehjs/src/lib/core/util/spatial.ts b/bokehjs/src/lib/core/util/spatial.ts
index 5bfa292587..7f093d518a 100644
--- a/bokehjs/src/lib/core/util/spatial.ts
+++ b/bokehjs/src/lib/core/util/spatial.ts
@@ -118,10 +118,10 @@ export class SpatialIndex {

   protected _normalize(rect: Rect): Rect {
     let {x0, y0, x1, y1} = rect
-    if (x0 > x1) {
+    if ((x0 > x1) && isFinite(x0 + x1)) {
       [x0, x1] = [x1, x0]
     }
-    if (y0 > y1) {
+    if ((y0 > y1) && isFinite(y0 + y1)) {
       [y0, y1] = [y1, y0]
     }
     return {x0, y0, x1, y1}
```
if that's what makes sense

Reply:
That makes sense. `_normalize()` is unfortunately necessary for the time being. It was easier to normalize in one place than trying to make sure all invocations provide coordinates in the correct order.